pkg/controller/infrastructure/infraflow/delete.go

```diff
@@ -72,14 +79,17 @@ func (fctx *FlowContext) buildDeleteGraph() *flow.Graph {
 
 	_ = fctx.AddTask(g, "delete share network",
 		fctx.deleteShareNetwork,
-		shared.Timeout(defaultTimeout), shared.Dependencies(recoverSubnetID))
+		shared.Timeout(defaultTimeout), shared.Dependencies(recoverSubnetID),
+		shared.DoIf(needToDeleteSubnet),
+	)
 	deleteRouterInterface := fctx.AddTask(g, "delete router interface",
 		fctx.deleteRouterInterface,
+		shared.DoIf(needToDeleteSubnet || needToDeleteRouter),
 		shared.Timeout(defaultTimeout), shared.Dependencies(recoverRouterID, recoverSubnetID, k8sRoutes))
 	// subnet deletion only needed if network is given by spec
 	_ = fctx.AddTask(g, "delete subnet",
 		fctx.deleteSubnet,
-		shared.DoIf(!needToDeleteNetwork), shared.Timeout(defaultTimeout), shared.Dependencies(deleteRouterInterface, k8sLoadBalancers))
+		shared.DoIf(!needToDeleteNetwork && needToDeleteSubnet), shared.Timeout(defaultTimeout), shared.Dependencies(deleteRouterInterface, k8sLoadBalancers))
```

Comment:
```suggestion
                // Delete the subnet only if the network is not being deleted already since it will be deleted anyways in that case.
		shared.DoIf(needToDeleteSubnet && !needToDeleteNetwork ), shared.Timeout(defaultTimeout), shared.Dependencies(deleteRouterInterface, k8sLoadBalancers))
```